packages/code-studio/vite.config.ts

```diff
@@ -0,0 +1,123 @@
+/* eslint-disable import/no-extraneous-dependencies */
+import { defineConfig, loadEnv } from 'vite';
+import react from '@vitejs/plugin-react';
+import path from 'path';
+import dns from 'dns';
+
+// Open on localhost instead of 127.0.0.1 for Node < 17
+// https://github.com/vitejs/vite/issues/9195
+dns.setDefaultResultOrder('verbatim');
+
+// https://vitejs.dev/config/
+export default defineConfig(({ mode }) => {
+  const env = loadEnv(mode, process.cwd(), '');
+
+  // https://github.com/vitejs/vite/issues/3105#issuecomment-939703781
+  const htmlPlugin = () => ({
+    name: 'html-transform',
+    transformIndexHtml: {
+      enforce: 'pre' as const,
+      transform(html: string) {
+        return html.replace(/#(.*?)#/g, (_, p1) => env[p1]);
+      },
+    },
+  });
+
+  const packagesDir = path.resolve(__dirname, '..');
+
+  // These are paths which should be proxied to the core server
+  // https://vitejs.dev/config/server-options.html#server-proxy
+  const proxy = {};
+  const proxyPaths = [
+    env.VITE_CORE_API_URL,
+    env.VITE_NOTEBOOKS_URL,
+    env.VITE_LAYOUTS_URL,
+  ];
+
+  if (env.VITE_PROXY_URL) {
+    proxyPaths.forEach(p => {
+      proxy[p] = {
+        target: env.VITE_PROXY_URL,
+        changeOrigin: true,
+      };
+    });
+  }
+
+  return {
+    root: 'src',
+    publicDir: path.resolve(__dirname, 'public'),
+    envDir: path.resolve(__dirname),
+    server: {
+      port: Number.parseInt(env.PORT, 10) ?? 4000,
+      open: true,
+      proxy,
+    },
+    resolve: {
+      dedupe: ['react', 'react-redux', 'redux'],
```

Comment:
dedupe ehhhh? Do we need to ensure our versions still match across package.jsons?

Reply:
I'm not sure we need the dedupe here actually (I added it when trying to fix an issue that was a different part of the config), but I figure it doesn't hurt.

[Vite docs](https://vitejs.dev/config/shared-options.html#resolve-dedupe) says it will always resolve these packages from the project root. So if for some reason our versions got out of sync and the hoisted version of React was not the version used in code-studio, we would still resolve to the version in `code-studio/node_modules/react`. Or if that doesn't exist then it would go to `<root>/node_modules/react`, but if it does not exist in code-studio node_modules, then the hoisted version satisfies what code-studio specified